File: api/handler/aluno.go
package handler

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"

	"github.com/CezarGarrido/cuco_robots/api/driver"
	entities "github.com/CezarGarrido/cuco_robots/api/entities"
	repo "github.com/CezarGarrido/cuco_robots/api/repository"
	"github.com/CezarGarrido/cuco_robots/crawler"
	"github.com/dgrijalva/jwt-go"
	"github.com/google/uuid"
)

var jwtKey = []byte("aplicativo_uems_dourados")

func NewAluno(db *driver.DB) *Aluno {
	return &Aluno{
		repo: repo.NewSQLAlunoRepo(db.SQL),
	}
}

type Aluno struct {
	repo repo.AlunoRepo
}

func (p *Aluno) Login(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	select {
	case <-ctx.Done():
		fmt.Fprint(os.Stderr, "request cancelled\n")
	case <-time.After(2 * time.Second):
		creds := &entities.Aluno{}
		err := json.NewDecoder(r.Body).Decode(creds)
		if err != nil {
			log.Println(err.Error())
			respondWithError(w, 500, "Login ou senha inválidos")
			return
		}
		isExists, err := p.repo.IsExiste(ctx, creds.Rgm, creds.Senha)
		if err != nil {
			log.Println(err.Error())
			respondWithError(w, 500, "Erro interno do sistema")
			return
		}
		if !isExists {
			client, err := crawler.NewClient(creds.Rgm, creds.Senha)
			if err != nil {
				log.Println(err.Error())
				respondWithError(w, 500, err.Error())
				return
			}
			aluno, err := client.FindAluno()
			if err != nil {
				log.Println(err.Error())
				respondWithError(w, 500, "Erro interno do sistema")
				return
			}
			hoje := time.Now()
			guid, err := uuid.NewRandom()
			if err != nil {
				log.Println(err.Error())
				respondWithError(w, 500, err.Error())
				return
			}
			newAluno := &entities.Aluno{
				Guid:  guid.String(),
				Nome:  aluno.Nome,
				Rgm:   creds.Rgm,
				Senha: creds.Senha,
				//Curso:           aluno.Curso,
				DataNascimento:  &aluno.DataNascimento,
				Sexo:            &aluno.Sexo,
				NomePai:         &aluno.NomePai,
				NomeMae:         &aluno.NomeMae,
				EstadoCivil:     &aluno.EstadoCivil,
				Nacionalidade:   &aluno.Nacionalidade,
				Naturalidade:    &aluno.Naturalidade,
				Fenotipo:        &aluno.Fenotipo,
				CPF:             &aluno.CPF,
				RG:              &aluno.RG,
				RGOrgaoEmissor:  &aluno.RGOrgaoEmissor,
				RGEstadoEmissor: &aluno.RGEstadoEmissor,
				RGDataEmissao:   &aluno.RGDataEmissao,
				CreatedAt:       &hoje,
			}
			contatos := make([]*entities.Contato, 0)
			for _, contato := range aluno.Contatos {
				newContato := &entities.Contato{
					Tipo:  contato.Tipo,
					Valor: &contato.Valor,
				}
				newContato.CreatedAt = &hoje
				contatos = append(contatos, newContato)
			}
			newAluno.Contatos = contatos
			enderecos := make([]*entities.Endereco, 0)
			for _, endereco := range aluno.Enderecos {
				newEndereco := &entities.Endereco{
					Logradouro:  &endereco.Logradouro,
					Numero:      &endereco.Numero,
					Complemento: &endereco.Complemento,
					Bairro:      &endereco.Bairro,
					CEP:         &endereco.CEP,
					Cidade:      &endereco.Cidade,
				}
				newEndereco.CreatedAt = &hoje
				enderecos = append(enderecos, newEndereco)
			}
			newAluno.Enderecos = enderecos
			cookie := client.GetCookies()[0]
			fmt.Println(cookie.Value, cookie.Name)
			newSessao := &entities.Sessao{
				QtdeLogin:   1,
				QtdeRequest: 1,
				CookieName:  cookie.Name,
				CookieValue: cookie.Value,
				CreatedAt:   hoje,
				UpdatedAt:   &hoje,
			}
			newAluno.Sessao = newSessao

			_, err = p.repo.Create(ctx, newAluno)
			if err != nil {
				log.Println(err.Error())
				respondWithError(w, 500, "Erro interno do sistema")
				return
			}
			//_, _ = client.Logout()
		}
		payload, err := p.repo.GetByLogin(ctx, creds.Rgm)
		if err != nil {
			log.Println(err.Error())
			respondWithError(w, 500, err.Error())
			return
		}
		if payload.Senha != creds.Senha {
			log.Println("Login ou senha inválidos", payload.Senha, creds.Senha)
			respondWithError(w, 500, "Senha inválida")
			return
		}
		//expirationTime := time.Now().Add(20 * time.Minute)
		claims := &entities.Claims{
			Aluno: *payload,
			/*StandardClaims: jwt.StandardClaims{
				//ExpiresAt: expirationTime.Unix(),
			},*/
		}
		token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
		tokenString, err := token.SignedString(jwtKey)
		if err != nil {
			log.Println(err.Error())
			respondWithError(w, 500, "Erro interno do sistema")
			return
		}

		respondwithJSON(w, 200, tokenString)
	}
}

// respondwithJSON write json response format
func respondwithJSON(w http.ResponseWriter, code int, payload interface{}) {
	response, _ := json.Marshal(payload)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	w.Write(response)
}

// respondwithError return error message
func respondWithError(w http.ResponseWriter, code int, msg string) {
	respondwithJSON(w, code, map[string]string{"message": msg})
}

File: app/lib/entities/aluno.dart
import 'contato.dart';
import 'endereco.dart';



class Aluno {
  int id;
  String guid;
  String nome;
  String rgm;
  String senha;
  String curso;
  String dataNascimento;
  String sexo;
  String nomePai;
  String nomeMae;
  String estadoCivil;
  String nacionalidade;
  String naturalidade;
  String fenotipo;
  String cpf;
  String rg;
  String rgOrgaoEmissor;
  String rgEstadoEmissor;
  String rgDataEmissao;
  List<Contato> contatos;
  List<Endereco> enderecos;
  String createdAt;
  String updatedAt;

  Aluno(
      {this.id,
      this.guid,
      this.nome,
      this.rgm,
      this.senha,
      this.curso,
      this.dataNascimento,
      this.sexo,
      this.nomePai,
      this.nomeMae,
      this.estadoCivil,
      this.nacionalidade,
      this.naturalidade,
      this.fenotipo,
      this.cpf,
      this.rg,
      this.rgOrgaoEmissor,
      this.rgEstadoEmissor,
      this.rgDataEmissao,
      this.contatos,
      this.enderecos,
      this.createdAt,
      this.updatedAt});

  Aluno.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    guid = json['guid'];
    nome = json['nome'];
    rgm = json['rgm'];
    senha = json['senha'];
    curso = json['curso'];
    dataNascimento = json['data_nascimento'];
    sexo = json['sexo'];
    nomePai = json['nome_pai'];
    nomeMae = json['nome_mae'];
    estadoCivil = json['estado_civil'];
    nacionalidade = json['nacionalidade'];
    naturalidade = json['naturalidade'];
    fenotipo = json['fenotipo'];
    cpf = json['cpf'];
    rg = json['rg'];
    rgOrgaoEmissor = json['rg_orgao_emissor'];
    rgEstadoEmissor = json['rg_estado_emissor'];
    rgDataEmissao = json['rg_data_emissao '];
    if (json['contatos'] != null) {
      contatos = new List<Contato>();
      json['contatos'].forEach((v) {
        contatos.add(new Contato.fromJson(v));
      });
    }
    if (json['enderecos'] != null) {
      enderecos = new List<Endereco>();
      json['enderecos'].forEach((v) {
        enderecos.add(new Endereco.fromJson(v));
      });
    }
    createdAt = json['created_at'];
    updatedAt = json['updated_at'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this.id;
    data['guid'] = this.guid;
    data['nome'] = this.nome;
    data['rgm'] = this.rgm;
    data['senha'] = this.senha;
    data['curso'] = this.curso;
    data['data_nascimento'] = this.dataNascimento;
    data['sexo'] = this.sexo;
    data['nome_pai'] = this.nomePai;
    data['nome_mae'] = this.nomeMae;
    data['estado_civil'] = this.estadoCivil;
    data['nacionalidade'] = this.nacionalidade;
    data['naturalidade'] = this.naturalidade;
    data['fenotipo'] = this.fenotipo;
    data['cpf'] = this.cpf;
    data['rg'] = this.rg;
    data['rg_orgao_emissor'] = this.rgOrgaoEmissor;
    data['rg_estado_emissor'] = this.rgEstadoEmissor;
    data['rg_data_emissao '] = this.rgDataEmissao;
    if (this.contatos != null) {
      data['contatos'] = this.contatos.map((v) => v.toJson()).toList();
    }
    if (this.enderecos != null) {
      data['enderecos'] = this.enderecos.map((v) => v.toJson()).toList();
    }
    data['created_at'] = this.createdAt;
    data['updated_at'] = this.updatedAt;
    return data;
  }
}